zmq.hpp

```diff
@@ -634,7 +634,146 @@ enum class socket_type : int
 #endif
     pair = ZMQ_PAIR
 };
-#endif
+
+struct send_result
+{
+    size_t bytes_sent;
+    bool success;
+};
+
+struct recv_result
+{
+    size_t bytes_received;
+    bool success;
+};
+
+enum class send_flags : int
+{
+    none = 0,
+    dontwait = ZMQ_DONTWAIT,
+    sndmore = ZMQ_SNDMORE
+};
+
+constexpr send_flags operator|(send_flags a, send_flags b) noexcept
+{
+    return static_cast<send_flags>(static_cast<int>(a) | static_cast<int>(b));
+}
+
+enum class recv_flags : int
+{
+    none = 0,
+    dontwait = ZMQ_DONTWAIT
+};
+
+constexpr recv_flags operator|(recv_flags a, recv_flags b) noexcept
+{
+    return static_cast<recv_flags>(static_cast<int>(a) | static_cast<int>(b));
+}
+
+struct from_handle_t
+{
+    struct _private {}; // disabling use other than with from_handle
+    constexpr explicit from_handle_t(_private /*p*/) noexcept {}
+};
+
+constexpr from_handle_t from_handle = from_handle_t{from_handle_t::_private{}};
+
+// A non-owning nullable reference to a socket.
+// The reference is invalidated on socket close or destruction.
+class socket_ref
```

Comment:
I am a bit confused on the relationship of socket_ref and socket. So now socket_ref has new send/recv variants, which are not present on socket. This is somewhat confusing IMO.

However, I am not sure what I would expect from the API of a socket_ref. Wouldn't one want to have the same operations as on socket?

That would lead to quite a lot of duplication though.

Wouldn't it be easier and implicitly consistent to use the same class for an owning and a non-owning socket, or at least the same base class (or base class template)?

Reply:
Yes, the API is supposed to be mostly the same except for the `close` function. `socket_t` also has the burden of pre-C++11 support (is dropping that support on the horizon?), but I will see if I can refactor the common functionality into a base class.